### PR TITLE
SONAR-7923 Create WS api/qualitygates/get_by_project

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

**Comment by stas-vilchik** [comment]
General comment: many unrelated changes in the second commit, it's hard to review.

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/GetByProjectAction.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    return qualityGateId.isPresent()
+      ? new QualityGateData(Optional.ofNullable(dbClient.qualityGateDao().selectById(dbSession, qualityGateId.get())), false)
+      : new QualityGateData(Optional.ofNullable(qualityGates.getDefault()), true);
+  }
+
+  private static class QualityGateData {
+    private final Optional<QualityGateDto> qualityGate;
+    private final boolean isDefault;
+
+    private QualityGateData(Optional<QualityGateDto> qualityGate, boolean isDefault) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/GetByProjectAction.java]: MINOR Specify a "QualityGateDto" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/CopyAction.java [hunk 2]
```diff
@@ -42,22 +43,22 @@ public void define(WebService.NewController controller) {
...
+    action.createParam(QualityGatesWsParameters.PARAM_NAME)
       .setDescription("The name of the quality gate to create")
       .setRequired(true)
       .setExampleValue("My Quality Gate");
   }
 
   @Override
   public void handle(Request request, Response response) {
-    QualityGateDto newQualityGate = qualityGates.copy(QGatesWs.parseId(request, QGatesWs.PARAM_ID), request.mandatoryParam(QGatesWs.PARAM_NAME));
+    QualityGateDto newQualityGate = qualityGates.copy(QualityGatesWs.parseId(request, QualityGatesWsParameters.PARAM_ID), request.mandatoryParam(QualityGatesWsParameters.PARAM_NAME));
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/CopyAction.java]: MINOR Split this 183 characters long line (which is greater than 180 authorized). [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MINOR 2 minor

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/GetByProjectAction.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    action.createParam(PARAM_PROJECT_KEY)
+      .setDescription("Project key")
+      .setExampleValue(KEY_PROJECT_EXAMPLE_001);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      ComponentDto project = componentFinder.getByUuidOrKey(dbSession, request.param(PARAM_PROJECT_ID), request.param(PARAM_PROJECT_KEY), ParamNames.PROJECT_ID_AND_KEY);
```
Comment (stas-vilchik) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/GetByProjectAction.java]: check that user has project "Browse" or "Administer" permission or global "Administer" permission

Comment (stas-vilchik) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/GetByProjectAction.java]: and add tests for these cases

File: server/sonar-server/src/test/java/org/sonar/server/qualitygate/ws/GetByProjectActionTest.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+    String result = ws.newRequest().setParam(PARAM_PROJECT_ID, project.uuid()).execute().getInput();
+
+    assertJson(result)
+      .ignoreFields("id")
+      .isSimilarTo(getClass().getResource("get_by_project-example.json"));
+  }
+
+  @Test
+  public void empty_response() {
```
Comment (stas-vilchik) [inline server/sonar-server/src/test/java/org/sonar/server/qualitygate/ws/GetByProjectActionTest.java]: test when there is a quality gate, but it's not default one and not associated to a project

File: server/sonar-server/src/test/java/org/sonar/server/qualitygate/ws/GetByProjectActionTest.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    }
+  }
+
+  private QualityGateDto insertQualityGate(String name) {
+    QualityGateDto qualityGate = dbClient.qualityGateDao().insert(dbSession, new QualityGateDto().setName(name));
+    db.commit();
+    return qualityGate;
+  }
+
+  private void associateProjectToQualityGate(long componentId, long qualityGateId) {
```
Comment (stas-vilchik) [inline server/sonar-server/src/test/java/org/sonar/server/qualitygate/ws/GetByProjectActionTest.java]: Do you think it's possible to reuse `org.sonar.server.qualitygate.QualityGates#associateProject`?

Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/qualitygate/ws/GetByProjectActionTest.java]: it's theorically possible but we want to have as few dependencies as possible. We set the data and QualityGates is a business oriented component. Possible in this specific case (as it's a dependency) but I prefer not to

**Comment by stas-vilchik** [comment]
LGTM

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Pull request merged**